server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2193,7 +2205,10 @@ public void performRecoveryRestart() throws IOException {
         assert Thread.holdsLock(mutex) == false : "restart recovery under mutex";
         synchronized (engineMutex) {
             assert refreshListeners.pendingCount() == 0 : "we can't restart with pending listeners";
-            IOUtils.close(currentEngineReference.getAndSet(null));
+            // TODO Make it async
+            var future = new PlainActionFuture<Void>();
+            Engine.close(currentEngineReference.getAndSet(null), future);
+            FutureUtils.get(future);
```

Comment:
I don't think this is any better than a `noop()` listener unfortunately. We need to move whatever executes after the close call so that it's called when the listener completes.

Reply:
Yeah, I've looked at that and I'm not sure if that is possible. The code path there is very synchronous. Specifically this method is called via `RecoveryTarget`, `RecoveriesCollection`, `PeerRecoveryTargetService` and `RecoveryResponseHandler` while holding and acquiring locks. I think `resetEngineToGlobalCheckpoint` where we have to close engines synchronously has similar issues.

It seems that making `Engine#close` purely async is way more complicated that we originally thought.

Reply:
The recovery subsystem is fundamentally quite async, it shouldn't be too hard to push that asyncness down from there to here.

`resetEngineToGlobalCheckpoint` looks a little harder but not impossible.

Reply:
@DaveCTurner I've extracted making `resetEngineToGlobalCheckpoint` and `performRecoveryRestart` async into 111582. This required quite a few changes including converting `engineMutex` to `ReentrantLock`